### PR TITLE
[Better Customer Selection] Older WooCommerce plugin clients

### LINKED ISSUE #10308: [Better Customer Selection] Search Fields UI (WooCommerce Versions <8.0.0)
Name - Email - Username

### PR DESCRIPTION
**Pull request by toupper**
Closes: #10308 and #10309 

## Description

With this PR we implement a better customer selector for older WooCommerce plugin clients (< 8.0.0). As those clients don't have the empty fields filter feature in the API request we cannot show a list of items, but only search/add new customer. Furthermore, given that we cannot search through all fields either, we add some filters for the search: name (the only one we had before), username, and email. Read more here pdfdoF-3mO-p2

## Changes
- Check for advanced search eligibility in `CustomerSelectorViewModel`
- Implement UI for Older clients: Default empty state with a button, and filters.
- Implement business logic for filters.
- Tweak the search `CustomerAction`: we don't return customers in the completion block as they are synced from the DB anyways.
- Load also a light version of clients when searching in older clients as we do when retrieving the full customer list. Leave the production version unchanged (trigger individual requests for each customer)
- Unit tests

## Testing instructions

- Enable the `betterCustomerSelectionInOrder` feature flag in `DefaultFeatureFlagService`
- With a store having an WooCommerce plugin installed and active and version <8.0.0

### Check that the search by filter works correctly

1. Go to Orders
2. Tap on + to create a new order
3. Tap on + Add Customer Details
4. See that you don't see the customer list, but the filters and an empty state. Search for a client, for instance by the email. That way you can make sure that the filters work fine. (See video)
5. Select the customer
6. See that it's added to the order

https://github.com/woocommerce/woocommerce-ios/assets/1864060/cdea6474-2ba1-4ecd-9ac0-54319572f742

### Check that adding a customer manually works correctly

1. Go to Orders
2. Tap on + to create a new order
3. Tap on + Add Customer Details
4. Tap on Add details manually
5. Add details manually and tap on Done
6. See that the customer details are added to the order

https://github.com/woocommerce/woocommerce-ios/assets/1864060/73af9a84-abad-4625-86a4-5bff80165112

## Screenshots

---
- [X] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

### CODE REVIEW

File: WooCommerce/Classes/ViewRelated/Orders/Order Creation/CustomerSection/CustomerSelectorViewModel.swift [hunk 1]
```diff
@@ -15,6 +15,20 @@ final class CustomerSelectorViewModel {
...
+                return completion(false)
+            }
+
+            let isCustomerAdvanceSearchSupportedByWCPlugin = VersionHelpers.isVersionSupported(version: wcPlugin.version,
+                                                                               minimumRequired: Constants.wcPluginMinimumVersion)
+            completion(isCustomerAdvanceSearchSupportedByWCPlugin)
+        }
+        stores.dispatch(action)
+    }
+
```
Comment (iamgabrielma) [inline WooCommerce/Classes/ViewRelated/Orders/Order Creation/CustomerSection/CustomerSelectorViewModel.swift:31]: Neat 💯

File: WooCommerce/WooCommerceTests/ViewRelated/Orders/Order Creation/CustomerSection/CustomerSelectorViewModelTests.swift [hunk 2]
```diff
@@ -14,6 +14,127 @@ final class CustomerSelectorViewModelTests: XCTestCase {
...
+        waitForExpectation { expectation in
+            viewModel.isEligibleForAdvancedSearch() { isEligible in
+                result = isEligible
+
+                expectation.fulfill()
+            }
+        }
+
+        // Then
+        XCTAssertFalse(result)
```
Comment (iamgabrielma) [inline WooCommerce/WooCommerceTests/ViewRelated/Orders/Order Creation/CustomerSection/CustomerSelectorViewModelTests.swift]: Just for reading clarity, what do you think if we switch the variable names here (and the other tests)? 

```suggestion
        var isEligible = true
        waitForExpectation { expectation in
            viewModel.isEligibleForAdvancedSearch() { result in
                isEligible = result

                expectation.fulfill()
            }
        }

        // Then
        XCTAssertFalse(isEligible)
```

**Review by iamgabrielma — APPROVED** [review]
LGTM and works well! 🚢

File: WooCommerce/Classes/ViewRelated/Orders/Order Creation/CustomerSection/CustomerSelectorViewModel.swift (hunk 1 above)
Comment (toupper) [reply WooCommerce/Classes/ViewRelated/Orders/Order Creation/CustomerSection/CustomerSelectorViewModel.swift:31]: Thanks!

File: WooCommerce/WooCommerceTests/ViewRelated/Orders/Order Creation/CustomerSection/CustomerSelectorViewModelTests.swift (hunk 2 above)
Comment (toupper) [reply WooCommerce/WooCommerceTests/ViewRelated/Orders/Order Creation/CustomerSection/CustomerSelectorViewModelTests.swift]: Good idea! Done in adef247b2253f4a0818e4fc9af42b52d1bb69b45

**Pull request merged**